# SOAL 1 - Data Transformation and Analysis Case
Christina Prilla Rosaria Ardyanti

## Import Library

In [2]:
#Melakukan import library yang dibutuhkan untuk data transformation dan analysis
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from datetime import datetime, timedelta

#Mengatur tampilan output dataframe pandas agar memudahkan analisis
pd.set_option('display.max_columns', None)          #set kolom jadi maksimal
pd.set_option('display.max_rows', None)             #set baris jadi maksimal
pd.options.display.float_format = '{:.2f}'.format   #set angka desimal 2 angka di belakang koma

## Data Understanding and Transformation

In [30]:
#Meload dataset dan menampilkan jumlah kolom dan baris pada dataset
df = pd.read_csv("ecommerce-session-bigquery.csv")
print('This dataset has %d rows dan %d columns.\n' % df.shape)
df.tail(20)

This dataset has 10000 rows dan 32 columns.



fullVisitorId channelGrouping    time        country  \
9980  7589075555607046758          Direct   33809         Taiwan   
9981  5864714452824654964        Referral  353975  United States   
9982  3519310386008495720          Direct  144343  United States   
9983  5684903298881626743  Organic Search  262769  United States   
9984  3570683077223774672          Direct   15574  United States   
9985  3137253471555883779        Referral  959947  United States   
9986  1054123144908181920          Social   13366   South Africa   
9987  6219236983626241836  Organic Search  278915  United States   
9988  6442689868405596515          Direct    4273          Spain   
9989  9795291451355594241        Referral   58512  United States   
9990  8160942479431419810  Organic Search    8679  United States   
9991  9622176459647341592          Direct       0        Germany   
9992   262583177454569381     Paid Search  177116  United States   
9993  2103516610461663217  Organic Search   50737  United States   
9994  4820449943911812081  Organic Search   49957          Japan   
9995  8518740059623315737        Referral   83144  United States   
9996  1592171014215970461  Organic Search  302378       Colombia   
9997  6776942687101397292      Affiliates   41387         Israel   
9998  3008082643543388126  Organic Search   28613         Canada   
9999  4307804717807925609  Organic Search  407285  United States   

                               city  totalTransactionRevenue  transactions  \
9980                      (not set)                      NaN           NaN   
9981                       New York                      NaN           NaN   
9982                       New York                      NaN           NaN   
9983                  Mountain View                      NaN           NaN   
9984                       New York                      NaN           NaN   
9985                       New York                      NaN           NaN   
9986  not available in demo dataset                      NaN           NaN   
9987                       New York                      NaN           NaN   
9988  not available in demo dataset                      NaN           NaN   
9989                    Los Angeles                      NaN           NaN   
9990                         Austin                      NaN           NaN   
9991  not available in demo dataset                      NaN           NaN   
9992  not available in demo dataset                      NaN           NaN   
9993                    Los Angeles                      NaN           NaN   
9994  not available in demo dataset                      NaN           NaN   
9995                      Ann Arbor                      NaN           NaN   
9996                         Bogota                      NaN           NaN   
9997  not available in demo dataset                      NaN           NaN   
9998                       Montreal                      NaN           NaN   
9999  not available in demo dataset                      NaN           NaN   

      timeOnSite  pageviews  sessionQualityDim      date     visitId  type  \
9980       64.00       4.00                NaN  20160815  1471249597  PAGE   
9981      446.00      25.00                NaN  20161002  1475458757  PAGE   
9982      173.00      13.00                NaN  20160822  1471889443  PAGE   
9983      643.00       7.00                NaN  20160912  1473695425  PAGE   
9984       16.00       2.00                NaN  20160908  1473391485  PAGE   
9985     1131.00      52.00                NaN  20160812  1471027456  PAGE   
9986       35.00       3.00                NaN  20161005  1475678729  PAGE   
9987      337.00      14.00                NaN  20160806  1470479897  PAGE   
9988        4.00       2.00                NaN  20160824  1472044349  PAGE   
9989      193.00      11.00                NaN  20160818  1471546034  PAGE   
9990     3184.00      80.00                NaN  20160802  1470167099  PAGE   
9991    

In [31]:
#Menampilkan informasi tipe data dan jumlah baris yang terisi dari tiap kolom/feature
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 32 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   fullVisitorId            10000 non-null  uint64 
 1   channelGrouping          10000 non-null  object 
 2   time                     10000 non-null  int64  
 3   country                  10000 non-null  object 
 4   city                     10000 non-null  object 
 5   totalTransactionRevenue  619 non-null    float64
 6   transactions             628 non-null    float64
 7   timeOnSite               9713 non-null   float64
 8   pageviews                9999 non-null   float64
 9   sessionQualityDim        19 non-null     float64
 10  date                     10000 non-null  int64  
 11  visitId                  10000 non-null  int64  
 12  type                     10000 non-null  object 
 13  productRefundAmount      0 non-null      float64
 14  productQuantity        

Perlu ada beberapa perbaikan pada dataset ini, seperti tipe data dan format tanggal, lalu menganalisis isi kolom yang kosong.

#### Distribusi Data Kategorikal

In [32]:
#Memilih kolom dengan tipe data string/object lalu menampilkan distribusi datanya
cat_column = df.select_dtypes(include=['object'])
cat_column.describe().transpose()

count unique                            top   freq
channelGrouping    10000      8                 Organic Search   4539
country            10000    106                  United States   6880
city               10000    171  not available in demo dataset   4668
type               10000      2                           PAGE   9745
productSKU         10000     76                 GGOEGBJC019999    911
v2ProductName      10000     84    Electronics Accessory Pouch    906
v2ProductCategory  10000     11                     Home/Bags/   5134
productVariant     10000      1                      (not set)  10000
currencyCode        9614      1                            USD   9614
pageTitle           9908      3                           Bags   7645
pagePathLevel1     10000      3              /google+redesign/   9861

#### Distribusi Data Numerik

In [33]:
#Memilih kolom dengan tipe data int dan float lalu menampilkan distribusi datanya
num_column = df.select_dtypes(include=['int64', 'float64'])
num_column.describe().transpose()

count          mean          std           min  \
time                    10000.00     324846.22    564722.10          0.00   
totalTransactionRevenue   619.00  210363537.96 565602597.82    3530000.00   
transactions              628.00          1.05         0.29          1.00   
timeOnSite               9713.00        720.12       968.00          1.00   
pageviews                9999.00         22.14        26.63          1.00   
sessionQualityDim          19.00         50.21        32.48          1.00   
date                    10000.00   20160965.76       906.76   20160801.00   
visitId                 10000.00 1473168925.20   2518655.22 1470035833.00   
productRefundAmount         0.00           NaN          NaN           NaN   
productQuantity            45.00          3.44         6.47          1.00   
productPrice            10000.00   15855650.00  20953393.94          0.00   
productRevenue              0.00           NaN          NaN           NaN   
itemQuantity                0.00           NaN          NaN           NaN   
itemRevenue                 0.00           NaN          NaN           NaN   
transactionRevenue          0.00           NaN          NaN           NaN   
transactionId               0.00           NaN          NaN           NaN   
searchKeyword               0.00           NaN          NaN           NaN   
eCommerceAction_type    10000.00          0.07         0.36          0.00   
eCommerceAction_step    10000.00          1.00         0.00          1.00   
eCommerceAction_option      0.00           NaN          NaN           NaN   

                                  25%           50%           75%  \
time                         42524.00     135718.00     354254.50   
totalTransactionRevenue   55080000.00   95420000.00  186660000.00   
transactions                     1.00          1.00          1.00   
timeOnSite                     164.00        383.00        898.00   
pageviews                        7.00         14.00         27.00   
sessionQualityDim               19.50         62.00         79.00   
date                      20160818.00   20160904.00   20160921.00   
visitId                 1471555271.00 1473018849.00 1474471268.75   
productRefundAmount               NaN           NaN           NaN   
productQuantity                  1.00          1.00          2.00   
productPrice               4990000.00    9990000.00   15990000.00   
productRevenue                    NaN           NaN           NaN   
itemQuantity                      NaN           NaN           NaN   
itemRevenue                       NaN           NaN           NaN   
transactionRevenue                NaN           NaN           NaN   
transactionId                     NaN           NaN           NaN   
searchKeyword                     NaN           NaN           NaN   
eCommerceAction_type             0.00          0.00          0.00   
eCommerceAction_step             1.00          1.00          1.00   
eCommerceAction_option            NaN           NaN           NaN   

                                  max  
time                      10894030.00  
totalTransactionRevenue 6831960000.00  
transactions                     6.00  
timeOnSite                   15020.00  
pageviews                      429.00  
sessionQualityDim               92.00  
date                      20170801.00  
visitId                 1501621677.00  
productRefundAmount               NaN  
productQuantity                 32.00  
productPrice              99990000.00  
productRevenue                    NaN  
itemQuantity                      NaN  
itemRevenue                       NaN  
transactionRevenue                NaN  
transactionId                     NaN  
searchKeyword                     NaN  
eCommerceAction_type             3.00  
eCommerceAction_step             1.00  
eCommerceAction_option            NaN

#### Kolom "date"

Pada dataset ini, kolom date memiliki format tanggal yang masih salah, sehingga format perlu diperbaiki.
Setelah diubah, terlihat bahwa dataset ini merupakan data yang diambil tepat 1 tahun, mulai dari tanggal 1 Agustus 2016 hingga 1 Agustus 2017.

In [34]:
df['date'] = pd.to_datetime(df['date'], format='%Y%m%d')
print("Data Tanggal Terlama : ", df['date'].min())
print("Data Tanggal Terbaru : ", df['date'].max())

Data Tanggal Terlama :  2016-08-01 00:00:00
Data Tanggal Terbaru :  2017-08-01 00:00:00


#### Kolom "time"

Kolom time masih memiliki format berupa second/detik, sehingga perlu diubah menjadi format datetime. Format tanggal dimulai berdasarkan tanggal awal pada dataset.

In [35]:
start_date = pd.to_datetime('2016-08-01')

df['time'] = start_date + pd.to_timedelta(df['time'], unit='s')


#### Kolom "pageViews"

Distribusi kolom "pageviews" memperlihatkan bahwa rata-rata customer mengunjungi sebanyak 22 halaman. Kunjungan halaman oleh customer terbanyak ada sebanyak 429.

In [38]:
df["pageviews"].describe()

count   9999.00
mean      22.14
std       26.63
min        1.00
25%        7.00
50%       14.00
75%       27.00
max      429.00
Name: pageviews, dtype: float64

#### Column "channelGrouping"

**ChatGPT Assistant**
Question 
```since i don't have any information about the schema of this dataset, there is one column named "channelGrouping" and the column contains this data
Organic Search    4539
Referral          2600
Direct            1872
Social             324
Paid Search        317
Affiliates         216
Display            131
(Other)              1

do you have any idea about the data?


Answer:
The column "channelGrouping" likely represents the different channels through which website or app visitors arrived. Here's what each category typically represents:

Organic Search: Visitors who arrived at the website through a search engine's organic (unpaid) search results. These visitors found the website by searching for specific keywords.

Referral: Visitors who were referred to the website by clicking on a link from another website. This category includes inbound traffic from external sources.

Direct: Visitors who accessed the website directly by typing the URL into their browser's address bar or using a bookmark. This category represents direct, non-referral traffic.

Social: Visitors who came to the website from social media platforms, such as Facebook, Twitter, or LinkedIn. This category includes traffic from social media posts and links.

Paid Search: Visitors who arrived at the website through paid search advertisements. This category includes traffic from pay-per-click (PPC) campaigns.

Affiliates: Visitors who came through affiliate marketing efforts. This category includes traffic generated by affiliates who promote the website.

Display: Visitors who arrived at the website through display advertising, such as banner ads on other websites or within apps.

(Other): This category might represent a catch-all for traffic sources that don't fit into the specific categories listed above. It could include less common or specialized sources.

The numbers next to each category likely represent the count of visitors or sessions that arrived through each channel. This information is valuable for analyzing the effectiveness of different marketing channels and understanding user acquisition.

Please note that the precise meanings of these categories can vary based on the analytics platform or data source, so it's important to refer to any available documentation or metadata associated with the dataset for a complete understanding.

Distribusi kolom "channelGrouping" memperlihatkan bahwa rata-rata customer mengunjungi halaman produk dengan melakukan pencarian secara langsung dengan mengetikkan keyword pada search engine atau yang melihat referensi produk dari website lain. Jumlah customer yang mengunjungi website dengan menggunakan link affiliate atau dari banner iklan paling sedikit. 

In [39]:
df['channelGrouping'].value_counts()

Organic Search    4539
Referral          2600
Direct            1872
Social             324
Paid Search        317
Affiliates         216
Display            131
(Other)              1
Name: channelGrouping, dtype: int64

#### Kolom "country" dan "city"

Total terdapat 106 negara yang melakukan transaksi pada e-commerce ini. Canada menjadi negara asal customer melakukan transaksi pada e-commerce.

In [91]:
print("Total jumlah negara : ", df['country'].nunique())
city = df.groupby('country')['city'].value_counts(ascending=False)
city

Total jumlah negara :  106


country               city                         
(not set)             (not set)                          11
                      not available in demo dataset       1
Algeria               (not set)                           1
                      not available in demo dataset       1
Argentina             Buenos Aires                       12
                      not available in demo dataset      11
Armenia               not available in demo dataset       1
Aruba                 not available in demo dataset       1
Australia             not available in demo dataset      43
                      Sydney                             36
                      Melbourne                          15
                      Brisbane                            1
Austria               not available in demo dataset      14
                      Vienna                              1
Bangladesh            (not set)                           3
Belarus               not available in demo data

#### Kolom "transaction"

Distribusi kolom "transaction" menampilkan bahwa kebanyakan transaksi hanya terjadi 1 kali, transaksi terbanyak yang dilakukan customer sejumlah 6 kali.

In [41]:
df['transactions'].describe()

count   628.00
mean      1.05
std       0.29
min       1.00
25%       1.00
50%       1.00
75%       1.00
max       6.00
Name: transactions, dtype: float64

#### Kolom "timeOnSite"

Distribusi kolom "timeOnSite" menampilkan bahwa kebanyakan customer mengunjugi website selama 12 menit, kunjungan paling lama yang dilakukan customer adalah selama 250 menit atau kurang lebih 4 jam.

In [42]:
df['timeOnSite (in minute)'] = df['timeOnSite'] / 60
df['timeOnSite (in minute)'].describe()

count   9713.00
mean      12.00
std       16.13
min        0.02
25%        2.73
50%        6.38
75%       14.97
max      250.33
Name: timeOnSite (in minute), dtype: float64

#### Kolom "sessionQualityDim"

In [43]:
df['sessionQualityDim'].describe()

count   19.00
mean    50.21
std     32.48
min      1.00
25%     19.50
50%     62.00
75%     79.00
max     92.00
Name: sessionQualityDim, dtype: float64

#### Kolom "productSKU" dan "v2ProductName"

**ChatGPT Assistant** 


Question:
```what is productSKU?```

Answer:
```"productSKU" typically refers to a product's Stock Keeping Unit (SKU). A SKU is a unique identifier or code assigned to a specific product or item in a retailer's inventory. SKUs are used for inventory management, tracking, and sales reporting. Here's what you need to know about "productSKU":

Unique Identifier: Each product in a retailer's inventory is assigned a distinct SKU, which serves as a unique identifier. This allows the retailer to track and manage individual products.

Inventory Management: SKUs are crucial for inventory management. They help retailers keep track of the quantity, location, and availability of each product in their stock.

Sales and Reporting: Retailers use SKUs to monitor sales, assess the popularity of products, and generate sales reports. By tracking sales by SKU, they can make informed decisions about restocking, pricing, and marketing.

Online Retail: In e-commerce, SKUs are often used to manage a wide variety of products. They help ensure that products are correctly listed on websites and that orders are fulfilled accurately.

Variants: SKUs can account for product variants, such as different colors, sizes, or styles. Each variant typically has its own SKU to avoid confusion.

Barcode: SKUs are often represented as barcodes, making it easy for retailers to scan products during inventory checks and at the point of sale.

Customization: The format of SKUs can vary between retailers, and they may choose to create SKUs in a way that aligns with their internal systems.

In summary, "productSKU" is a critical component of inventory management in the retail industry. It ensures that each product is uniquely identified, tracked, and reported on, contributing to efficient operations and a seamless shopping experience for customers.```

In [44]:
print("Total jumlah SKU : ", df['productSKU'].nunique())
city = df.groupby('productSKU')['v2ProductName'].value_counts().sort_values(ascending=False)
city

Total jumlah SKU :  76


productSKU      v2ProductName                                        
GGOEGBFC018799  Electronics Accessory Pouch                              906
GGOEGBJL013999  Google Canvas Tote Natural/Navy                          845
GGOEGBCR024399  Google Lunch Bag                                         835
GGOEGBJC019999  Collapsible Shopping Bag                                 690
GGOEGBJC014399  Google Tote Bag                                          675
GGOEGBJR018199  Reusable Shopping Bag                                    657
GGOEGBMJ013399  Sport Bag                                                522
GGOEGBJB022599  Large Zipper Top Tote Bag                                499
GGOEGBMC056599  Waterpoof Gear Bag                                       469
GGOEGBMB073799  Google Zipper-front Sports Bag                           452
GGOEGBMC056599  Waterproof Gear Bag                                      420
GGOEGBJR018199  Chevron Shopper                                          247
GGOEGB

#### Kolom "v2ProductCategory"

In [45]:
df['v2ProductCategory'].value_counts()

Home/Bags/                            5134
Home/Limited Supply/Bags/             2511
Home/Accessories/Fun/                 1785
Home/Accessories/Pet/                  301
Home/Fun/                              123
(not set)                               86
Home/Lifestyle/Fun/                     54
Home/Shop by Brand/YouTube/              2
${escCatTitle}                           2
Home/Apparel/Men's/Men's-T-Shirts/       1
Home/Bags/Backpacks/                     1
Name: v2ProductCategory, dtype: int64

#### Kolom "totalTransactionRevenue"

In [46]:
df[df["totalTransactionRevenue"].notnull()].tail()

fullVisitorId channelGrouping                time        country  \
9879  4825534357105727679        Referral 2016-08-01 04:45:18  United States   
9925  7166253532069552655         Display 2016-08-04 05:34:14  United States   
9938  5262771322194198136        Referral 2016-08-01 01:33:01  United States   
9958  4835031132385744572          Direct 2016-08-15 13:59:15  United States   
9968  2518447979646501002  Organic Search 2016-08-02 09:39:33  United States   

               city  totalTransactionRevenue  transactions  timeOnSite  \
9879  Mountain View             125100000.00          1.00      356.00   
9925  San Francisco             451700000.00          1.00     1187.00   
9938         Austin             150340000.00          1.00     1731.00   
9958      Palo Alto              62870000.00          1.00     1643.00   
9968       New York              78150000.00          1.00      705.00   

      pageviews  sessionQualityDim       date     visitId  type  \
9879      14.00                NaN 2016-08-09  1470762213  PAGE   
9925      41.00                NaN 2016-08-18  1471550764  PAGE   
9938      61.00                NaN 2016-08-22  1471906107  PAGE   
9958      32.00                NaN 2016-08-05  1470409970  PAGE   
9968      40.00                NaN 2016-08-20  1471740007  PAGE   

      productRefundAmount  productQuantity  productPrice  productRevenue  \
9879                  NaN              NaN       4990000             NaN   
9925                  NaN              NaN       4990000             NaN   
9938                  NaN              NaN       4990000             NaN   
9958                  NaN              NaN       4990000             NaN   
9968                  NaN              NaN       4990000             NaN   

          productSKU v2ProductName          v2ProductCategory productVariant  \
9879  GGOEGBMJ013399     Sport Bag  Home/Limited Supply/Bags/      (not set)   
9925  GGOEGBMJ013399     Sport Bag  Home/Limited Supply/Bags/      (not set)   
9938  GGOEGBMJ013399     Sport Bag                 Home/Bags/      (not set)   
9958  GGOEGBMJ013399     Sport Bag  Home/Limited Supply/Bags/      (not set)   
9968  GGOEGBMJ013399     Sport Bag                 Home/Bags/      (not set)   

     currencyCode  itemQuantity  itemRevenue  transactionRevenue  \
9879          USD           NaN          NaN                 NaN   
9925          USD           NaN          NaN                 NaN   
9938          USD           NaN          NaN                 NaN   
9958          NaN           NaN          NaN                 NaN   
9968          USD           NaN          NaN                 NaN   

      transactionId pageTitle  searchKeyword     pagePathLevel1  \
9879            NaN      Bags            NaN  /google+redesign/   
9925            NaN      Bags            NaN  /google+redesign/   
9938            NaN      Bags            NaN  /google+redesign/   
9958            NaN      Bags            NaN  /google+redesign/   
9968            NaN      Bags            NaN  /google+redesign/   

      eCommerceAction_type  eCommerceAction_step  eCommerceAction_option  \
9879                     0                     1                     NaN   
9925                     0                     1                     NaN   
9938                     0                     1                     NaN   
9958                     2                     1                     NaN   
9968                     0                     1                     NaN   

      timeOnSite (in minute)  
9879                    5.93  
9925                   19.78  
9938                   28.85  
9958                   27.38  
9968                   11.75

In [ ]:
df["totalTransactionRevenue"].describe()

#### Kolom "productRevenue", "itemRevenue", "transactionRevenue"

**ChatGPT Assistant**

Question:
```how to get productRevenue, itemRevenue, and transactionRevenue value?

is productRevenue the same value as productPrice or it can be reduced by a discount?

how about itemRevenue?```


Answer:

```To get the values of "productRevenue," "itemRevenue," and "transactionRevenue" columns in a pandas DataFrame:

python
Copy code
product_revenue_values = df['productRevenue']
item_revenue_values = df['itemRevenue']
transaction_revenue_values = df['transactionRevenue']
However, it's important to clarify the distinctions between these columns and their relationship with "productPrice" and discounts:

productRevenue:

"productRevenue" typically represents the total revenue generated from individual products sold in a transaction.
This value may include the price of the product before discounts, taxes, or other charges. It reflects the revenue from the sale of a product before any deductions.
productPrice:

"productPrice" usually represents the price of an individual product before any discounts are applied.
It is the base price at which a product is listed or sold before considering any discounts, promotions, or additional charges.
itemRevenue:

"itemRevenue" is similar to "productRevenue" but may refer to revenue generated from individual items or line items within a transaction.
It can represent the revenue from individual items within a transaction, including their respective prices before discounts.
transactionRevenue:

"transactionRevenue" represents the total revenue generated from an entire transaction or order.
It includes the cumulative revenue from all products or items within a single transaction.
In general, "productRevenue" and "itemRevenue" are related to the individual products or items sold within a transaction. These columns may not account for discounts or additional charges, as they focus on the revenue associated with the sale of the products or items themselves. Discounts, taxes, and other adjustments may be captured separately in the dataset or in other columns.

"productPrice" typically represents the base price of a single product before discounts are applied. If discounts are offered on the product, the difference between the "productPrice" and "productRevenue" or "itemRevenue" can indicate the amount of discount applied to each product.

It's essential to consult the dataset's documentation or data dictionary to fully understand how these columns are defined and used in your specific dataset.```

**ChatGPT Assistant**

Question:

```if i have data "totalTransactionRevenue" = 125100000 and productPrice = 4990000. can you calculate how much productRevenue, itemRevenue, and transactionRevenue```


Answer:
```Certainly, we can calculate "productRevenue," "itemRevenue," and "transactionRevenue" based on the provided data:

- `totalTransactionRevenue` = $125,100,000
- `productPrice` = $4,990,000

1. **Product Revenue (for a single product):**

   To calculate the "productRevenue" for a single product, you need to know the number of units sold. If you have the number of units sold (quantity), you can use the formula:

   ```
   Product Revenue = Quantity Sold × Product Price
   ```

   However, if you don't have the quantity of the product sold and want to calculate the "productRevenue" from the total transaction revenue and the product price, you can rearrange the formula as follows:

   ```
   Product Revenue = Total Transaction Revenue × (Product Price / Total Transaction Price)
   ```

   where `Total Transaction Price` is the total price for the transaction, and it's calculated based on the "productPrice" and "totalTransactionRevenue." In this case:

   ```
   Total Transaction Price = Total Transaction Revenue / Product Price
   ```

   Using the provided data:

   ```
   Total Transaction Price = $125,100,000 / $4,990,000 = 25
   ```

   Now, you can calculate the "productRevenue" for a single product:

   ```
   Product Revenue = $125,100,000 × ($4,990,000 / $125,100,000) = $4,990,000
   ```

2. **Item Revenue (for a single item):**

   "Item Revenue" is calculated in the same way as "Product Revenue" since they represent the revenue for individual products or items. Using the formula for "Product Revenue," the "Item Revenue" for a single item would also be $4,990,000.

3. **Transaction Revenue:**

   To calculate the "transactionRevenue" for the entire transaction, you would sum the "itemRevenue" for all items within the transaction. If you have the quantity of items sold for each product, you can calculate it as follows:

   ```
   Transaction Revenue = Σ (Item Revenue for all Items in the Transaction)
   ```

   If you have the quantity of items sold for each product, you can calculate the "transactionRevenue" based on the quantity of items sold.

   However, if you don't have the quantity of items sold, you may need more information to perform this calculation. Please provide the quantities sold for each product if available.```

**ChatGPT Assistant**

Question:
```but Total Transaction Price = $125,100,000 / $4,990,000 = 25 actually a decimal value, product quantity should be integers, so where's the rest fund? because when I calculate $125,100,000 % $4,990,000 = 350000```



Answer:

```I see your point. If "Total Transaction Price" is calculated as a decimal value (25.05) and the quantity of products sold should be integers, then you are correct that there would be a remainder or a discrepancy. In such cases, the remaining amount (in your example, $350,000) is not accounted for within the "Product Revenue" calculation, and it needs to be distributed or allocated.

One common way to handle this is to distribute the remainder evenly across the products or items in the transaction. Here's how you can adjust the calculations:

- **Product Revenue (for a single product):**
   - You can calculate the "Product Revenue" for a single product as before: 
     ```
     Product Revenue = Total Transaction Revenue × (Product Price / Total Transaction Price)
     ```
   - In your example, this would be $4,990,000.

- **Item Revenue (for a single item):**
   - If you have multiple items in the transaction, you would distribute the remaining $350,000 evenly across the items. Let's say you have three items:
     ```
     Item Revenue = Total Transaction Revenue × (Product Price / Total Transaction Price) / Number of Items
     ```
     - In your example, this would be ($4,990,000 / 25) / 3 = $66,400 per item.

- **Transaction Revenue:**
   - The "Transaction Revenue" would then be the sum of "Item Revenue" for all items:
     ```
     Transaction Revenue = Σ (Item Revenue for all Items in the Transaction)
     ```
     - In your example, it would be $66,400 + $66,400 + $66,400 = $199,200 for the entire transaction.

By distributing the remainder evenly across the items, you account for the fractional part of the "Total Transaction Price" while keeping the "Transaction Revenue" as an integer value. This approach ensures that the sum of "Item Revenue" for all items equals the "Transaction Revenue."```

**ChatGPT Assistant**

Question:
```so how to calculate the product quantity?```


Answer:

```To calculate the product quantity when you have the "Total Transaction Price," "Total Transaction Revenue," and "Product Price," you can use the following formula:

```
Product Quantity = Total Transaction Price / Product Price
```

This formula allows you to determine the quantity of a product that was sold within a transaction based on the total transaction price and the price per unit of the product. It provides the number of units of the product that could be sold at the given price, considering the total revenue of the transaction.

In your example, with a "Total Transaction Price" of 25.05 and a "Product Price" of $4,990,000:

```
Product Quantity = 25.05 / $4,990,000 ≈ 0.000005011 units
```

Please note that the result is a decimal value. If you require a whole number of units, you may need to round the result, which may result in rounding discrepancies due to the decimal values involved.

If the calculated quantity is indeed intended to be a whole number and you have information on any rounding or fractional quantities used in your dataset, you should consider any specific rules or conventions applied in your data.```

In [55]:
df['productRefundAmount'] = df['productQuantity'] * df['productPrice']

df['itemQuantity'] = ((df['totalTransactionRevenue'] - (df['totalTransactionRevenue'] % df['productPrice'])) / df['productPrice']) - df['productQuantity']

df['productRevenue'] = df['totalTransactionRevenue'] * df['productPrice'] / df['totalTransactionRevenue']

df['itemRevenue'] = (df['totalTransactionRevenue'] * df['productPrice'] / df['totalTransactionRevenue']) / df['itemQuantity']

df['transactionRevenue'] = df['itemQuantity'] * df['itemRevenue']

In [78]:
df['itemQuantity'].describe()

count    611.00
mean      33.64
std      102.01
min        0.00
25%        5.00
50%       12.00
75%       28.00
max     1490.00
Name: itemQuantity, dtype: float64

In [79]:
df['productRevenue'].describe()

count        619.00
mean    14324814.22
std     19622742.62
min            0.00
25%      3990000.00
50%      7990000.00
75%     13990000.00
max     79990000.00
Name: productRevenue, dtype: float64

In [80]:
df['itemRevenue'].describe()

count       611.00
mean           inf
std            NaN
min        2092.53
25%      144916.67
50%      518148.15
75%     2497500.00
max            inf
Name: itemRevenue, dtype: float64

In [81]:
df['transactionRevenue'].describe()

count        585.00
mean    12434735.04
std     16412933.21
min      1200000.00
25%      3990000.00
50%      6990000.00
75%     13990000.00
max     79990000.00
Name: transactionRevenue, dtype: float64

In [82]:
#Menyimpan dataset yang telah dibersihkan ke dalam file csv
df.to_csv("clean.csv")

## Exploratory Data Analysis